README.md
# Gotipath Stream Uploader

## Installation

Add the package to the `dependencies` section in `pubspec.yaml`:
- `gotipath_stream_uploader: ^1.0.1` (or latest release)

## Usage

Add the following import to the `.dart` file that will use **GotipathStreamUploader**

`import 'package:gotipath_stream_uploader/gotipath_stream_uploader.dart';`

### Example

```dart
  // ADD ENDPOINT and credential HERE
final String _endPoint = "https://apistream.gotipath.com/v1/";
final String _clientID = 'Your client ID';
final String _libraryID = 'Your library ID';
final String _apiKey = 'Your API key';
final String _videoID = 'Your video ID';

GotipathStreamUploader gotipathStreamUploader = GotipathStreamUploader();


// Chunk upload
gotipathStreamUploader
  ..endPoint = _endPoint
  ..clientID = _clientID
  ..libraryID = _libraryID
  ..apiKey= _apiKey
  ..videoID = _videoID
  ..file = fileToUpload
  ..onProgress = (double progress) {
  setState(() {
  _progress = progress.ceil();
  });
  }
  ..onError = (String message, int chunk, int attempts) {
  setState(() {
  _errorMessage = 'error 💥 🙀:\n'
  ' - Message: $message\n'
  ' - part: $chunk\n'
  ' - Attempts: $attempts';
  });
  }
  ..onSuccess = () {
  setState(() {
  _uploadComplete = true;
  });
  };



gotipathStreamUploader.createUpload();
```

## API

Although the API is a port of the original JS library, some options and properties differ slightly.

### `createUpload()`

Intializes the upload process. This method must be called after the `GotipathUploader` instance is created and all event handlers are set.

#### `GotipathStreamUploader` parameters:

##### Upload options

- `endPoint` <small>type: `string` (required if `endPointResolver` is `null`)</small>
- `clientID` <small>type: `string` (required)</small>
- `libraryID` <small>type: `string` (required)</small>
- `apiKey` <small>type: `string` (required)</small>
- `videoID` <small>type: `string` (required)</small>


  URL to upload the file to.

- `endPointResolver` <small>type: `Future<String>` (required if `endPoint` is `null`)</small>

  A `Future` that returns the URL as a `String`.

- `file` <small>type: [`File`](https://api.dart.dev/stable/2.10.3/dart-io/File-class.html) (required)</small>

  The file you'd like to upload.

- `headers` <small>type: `Map<String, String>`</small>

  A `Map` with any headers you'd like included with the `PUT` request for each chunk.

- `chunkSize` <small>type: `integer`, default:`5120`</small>

  The size in kb of the chunks to split the file into, with the exception of the final chunk which may be smaller. This parameter should be in multiples of 64.

- `attempts` <small>type: `integer`, default: `5`</small>

  The number of times to retry any given chunk.

- `delayBeforeRetry` <small>type: `integer`, default: `1`</small>

  The time in seconds to wait before attempting to upload a chunk again.

##### Event options

- `onAttempt` <small>`{ chunkNumber: Integer, chunkSize: Integer }`</small>

  Fired immediately before a chunk upload is attempted. `chunkNumber` is the number of the current chunk being attempted, and `chunkSize` is the size (in bytes) of that chunk.

- `onAttemptFailure` <small>`{ message: String, chunkNumber: Integer, attemptsLeft: Integer }`</small>

  Fired when an attempt to upload a chunk fails.

- `onError` <small>`{ message: String, chunk: Integer, attempts: Integer }`</small>

  Fired when a chunk has reached the max number of retries or the response code is fatal and implies that retries should not be attempted.

- `onOffline`

  Fired when the client has gone offline.

- `onOnline`

  Fired when the client has gone online.

- `onProgress` <small>`progress double [0..100]`</small>

  Fired continuously with incremental upload progress. This returns the current percentage of the file that's been uploaded.

- `onSuccess`

  Fired when the upload is finished successfully.

### GotipathStreamUploader Instance Methods

- `pause()`

  Pauses an upload after the current in-flight chunk is finished uploading.

- `resume()`

  Resumes an upload that was previously paused.

- `cancel()`

  Cancels the upload abruptly.



CHANGELOG.md
## [1.0.1] - 2023-11-27

* Flutter Package for uploading files to Gotipath Stream


test/gotipath_gploader_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:gotipath_uploader/gotipath_uploader.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/lib/main.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:flutter/material.dart';
import 'package:gotipath_stream_uploader/gotipath_stream_uploader.dart';
import 'package:image_picker/image_picker.dart';

import 'credential.dart';



void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'UpChunk Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Gotipath Stream Multipart'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({ this.title = ''});

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  // ADD ENDPOINT URL HERE


  final picker = ImagePicker();

  int _progress = 0;
  bool _uploadComplete = false;
  String _errorMessage = '';

  String fileToUpload = '';
  GotipathStreamUploader gotipathStreamUploader = GotipathStreamUploader();

  void _getFile() async {
    fileToUpload= '';
    final pickedFile = await picker.pickVideo(source: ImageSource.gallery);

    if (pickedFile == null) return;

    setState(() {
      fileToUpload = pickedFile.path;
    });

   final videoID=await createUploadUrlRequest(endPoint: endPoint,clientID: clientID,libraryID: libraryID,apiKey: apiKey);
   videoId=videoID;
  //  _uploadFile(File(pickedFile.path));
  }


  Future<String> createUploadUrlRequest(
      {String? endPoint,
        String? clientID,
        String? libraryID,
        String? apiKey}
      ) async {

    final String url = endPoint! + 'libraries/$libraryID/videos';
    final client = new http.Client();

    try{
      final response = await client.post(
        Uri.parse(url),
        headers: {'Accept': 'application/json', 'Content-type': 'application/json', "X-Auth-ClientId": clientID!, "X-Auth-LibraryId": libraryID!, "X-Auth-ApiKey": apiKey! },
        body: jsonEncode(<String, String>{
          'name': fileToUpload.split('/').last,
        }),
      );
      if (response.statusCode == 200) {

        return jsonDecode(response.body)['result']['id'];

      } else {
        return "";
      }
    }catch(e){
      print("error ${e}");
      return "";
    }

  }

  void _uploadFile(File fileToUpload) {
    _progress = 0;
    _uploadComplete = false;
    _errorMessage = '';

    gotipathStreamUploader
      ..endPoint = endPoint
      ..clientID = clientID
      ..libraryID = libraryID
      ..apiKey= apiKey
      ..videoID = videoId
      ..file = fileToUpload
      ..onProgress = (double progress) {
        setState(() {
          _progress = progress.ceil();
        });
      }
      ..onError = (String message, int chunk, int attempts) {
        setState(() {
          _errorMessage = 'UpChunk error 💥 🙀:\n'
              ' - Message: $message\n'
              ' - Chunk: $chunk\n'
              ' - Attempts: $attempts';
        });
      }
      ..onSuccess = () {
        setState(() {
          _uploadComplete = true;
        });
      };



    gotipathStreamUploader.createUpload();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          children: [
            if (!_uploadComplete)
              Text(
                'Uploaded: $_progress%',
                style: TextStyle(
                  fontSize: 24,
                  fontWeight: FontWeight.normal,
                ),
              ),

            if (_uploadComplete)
              Text(
                'Upload complete! 👋',
                style: TextStyle(
                  fontSize: 24,
                  color: Colors.green,
                  fontWeight: FontWeight.bold,
                ),
              ),

            if (_errorMessage.isNotEmpty)
              Text(
                '$_errorMessage%',
                style: TextStyle(
                  fontSize: 24,
                  fontWeight: FontWeight.normal,
                  color: Colors.red,
                ),
              ),

            SizedBox(height: 30),
            InkWell(
              onTap: (){
                _uploadFile(File(fileToUpload));
              },
              child: Container(
                width: 200,
                height: 50,
                color: Colors.grey,
                child: Center(
                  child: Text(
                    'Video upload',
                    style: TextStyle(
                      fontSize: 24,
                      fontWeight: FontWeight.normal,
                      color: Colors.white,
                    ),
                  ),
                ),
              ),
            ),
            SizedBox(height: 30),
            InkWell(
              onTap: (){
                // paused.value=true;
                gotipathStreamUploader.pause();

              },
              child: Container(
                width: 200,
                height: 50,
                color: Colors.blue,
                child: Center(
                  child: Text(
                    'Upload Pause',
                    style: TextStyle(
                      fontSize: 24,
                      fontWeight: FontWeight.normal,
                      color: Colors.white,
                    ),
                  ),
                ),
              ),
            ),
            SizedBox(height: 30),
            InkWell(
              onTap: (){
                // paused.value=false;
                gotipathStreamUploader.resume();
              //  GotipathStreamUploader.resume();
              },
              child: Container(
                width: 200,
                height: 50,
                color: Colors.blue,
                child: Center(
                  child: Text(
                    'Upload Resume',
                    style: TextStyle(
                      fontSize: 24,
                      fontWeight: FontWeight.normal,
                      color: Colors.white,
                    ),
                  ),
                ),
              ),
            ),

            SizedBox(height: 30),
            InkWell(
              onTap: (){
                // paused.value=false;
                gotipathStreamUploader.cancel();
                //  GotipathStreamUploader.resume();
              },
              child: Container(
                width: 200,
                height: 50,
                color: Colors.blue,
                child: Center(
                  child: Text(
                    'Upload Cancel',
                    style: TextStyle(
                      fontSize: 24,
                      fontWeight: FontWeight.normal,
                      color: Colors.white,
                    ),
                  ),
                ),
              ),
            ),

          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _getFile,
        tooltip: 'Get File',
        child: Icon(Icons.upload_file),
      ),
    );
  }
}


lib/src/multipart.dart
import 'dart:async';
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:mime/mime.dart';
import 'package:http/http.dart' as http;
import 'connection_status_singleton.dart';
import 'uploader_repository.dart';


class GotipathStreamUploader {
  /// HTTP response codes implying the PUT method has been successful
  final successfulChunkUploadCodes = const [200, 201, 202, 204, 308];

  /// HTTP response codes implying a chunk may be retried
  final temporaryErrorCodes = const [408, 502, 503, 504];

  String? endPoint;

  String? clientID;

  String? libraryID;

  String? apiKey;

  String? videoID;

  String? collectionID;

  String? upload_id;

  String? upload_key;

  List<Map<String,dynamic>> chunk_list=[];

  Future<String>? endPointResolver;
  File? file;
//  Map<String, String> headers = {};
  int chunkSize = 5120;
  int attempts = 5;
  int delayBeforeAttempt = 1;

  Stream<List<int>> _chunk = Stream.empty();
  int _chunkLength = 0;
  int _fileSize = 0;
  int _chunkCount = 0;
  int chunkByteSize = 0;
  String? _fileMimeType;
  Uri _endpointValue = Uri();
  int _totalChunks = 0;
  int _attemptCount = 0;
  bool _offline = false;
  bool _paused = false;
  bool _stopped = false;

  CancelToken? _currentCancelToken;

  bool _uploadFailed = false;

  void Function()? onOnline;
  void Function()? onOffline;
  void Function(int chunkNumber, int chunkSize)? onAttempt;
  void Function(String message, int chunkNumber, int attemptsLeft)? onAttemptFailure;
  void Function(String message, int chunk, int attempts)? onError;
  void Function()? onSuccess;
  void Function(double progress )? onProgress;

  GotipathStreamUploader();

   createUpload() => _internal();

  /// Internal constructor used by [createUpload]
  _internal() {


    _validateOptions();

    chunkByteSize = chunkSize * 1024;



    videoUploadRequest(endPoint: endPoint, clientID: clientID, libraryID: libraryID, apiKey: apiKey, filename: file!.path.split('/').last, videoID: videoID)
      .then((value) async {
        upload_id=value['uploadId'];
        upload_key=value['key'];
        _fileSize = await file!.length();
        _totalChunks =  (_fileSize / chunkByteSize).ceil();

        await _getMimeType();
      })
      .then((_) => _sendChunks());

    // restart sync when back online
    // trigger events when offline/back online
    ConnectionStatusSingleton connectionStatus = ConnectionStatusSingleton();


    connectionStatus.connectionChange.listen(_connectionChanged);
  }

  /// It pauses the upload, the [_chunk] currently being uploaded will finish first before pausing the next [_chunk]
   pause() => _paused = true;



  /// It resumes the upload for the next [_chunk]
  resume() {
    if (!_paused) return;

    _paused = false;
    _sendChunks();
  }

  cancel() async{
    _stopped = true;
    _uploadFailed = true;
    _currentCancelToken!.cancel(Exception('Upload cancelled by the user'));
    await abortUpload(endPoint: endPoint, clientID: clientID, libraryID: libraryID, apiKey: apiKey, upload_id: upload_id, upload_key: upload_key);

    if (onError != null)
      onError!(
        'Upload cancelled by the user.',
        _chunkCount,
        _attemptCount,
      );
  }

  /// It gets [file]'s mime type, if possible
  _getMimeType() async {
    try {
      _fileMimeType = lookupMimeType(file!.path);
    } catch (_) {
      _fileMimeType = null;
    }
  }

  /// It validates the passed options
  _validateOptions() {
    if (endPoint == null && endPointResolver == null)
      throw new Exception('either endPoint or endPointResolver must be defined');

    if (file == null)
      throw new Exception('file can''t be null');

    if (chunkSize <= 0 || chunkSize % 64 != 0)
      throw new Exception('chunkSize must be a positive number in multiples of 64');

    if (attempts <= 0)
      throw new Exception('retries must be a positive number');

    if (delayBeforeAttempt < 0)
      throw new Exception('delayBeforeAttempt must be a positive number');
  }




  /// Callback for [ConnectionStatusSingleton] to notify connection changes
  ///
  /// if the connection drops [_offline] is marked as true and upload us paused,
  /// if connection is restore [_offline] is marked as false and resumes the upload
  _connectionChanged(dynamic hasConnection) {
    if (hasConnection) {
      if (!_offline)
        return;

      _offline = false;

      if (onOnline != null) onOnline!();

      _sendChunks();
    }

    if (!hasConnection) {
      _offline = true;

      if (onOffline != null) onOffline!();
    }
  }

  /// Sends [_chunk] of the file with appropriate headers
  Future<Response> _sendChunk(String presignedUrl,int chunkLenght) async {

    var putHeaders =  {"Accept":"*/*","Content-Length": chunkLenght, "Content-Type": "binary/octet-stream"};

    // if (_fileMimeType != null){
    //   putHeaders.putIfAbsent(Headers.contentTypeHeader, () => _fileMimeType!);
    // }
    // headers.forEach((key, value) => putHeaders.putIfAbsent(key, () => value));

    if (onAttempt != null)
      onAttempt!(_chunkCount, _chunkLength);

    _currentCancelToken = CancelToken();

   // Response<Response> response =await http.put(Uri.parse(presignedUrl), body: _chunk, headers: {"Accept":"*/*","Content-Length": chunkByteSize.toString()});

    final response=await Dio().request(
      presignedUrl,
      options: Options(
          headers: putHeaders,
          method: "PUT",
          // contentType:'video/mp4',
          followRedirects: false,
          validateStatus: (status) {
            return true;
          }
      ),
      data: _chunk,
      onSendProgress: (int sent, int total) {
        if (onProgress != null) {
          final bytesSent = _chunkCount * chunkByteSize;
          final percentProgress = (bytesSent + sent) * 100.0 / _fileSize;

          if (percentProgress < 100.0)
            onProgress!(percentProgress);
        }
      },
      cancelToken: _currentCancelToken,
    );
    // returns future with http response
    return response;
  }

  /// Gets [_chunk] and [_chunkLength] for the portion of the file of x bytes corresponding to [chunkByteSize]
  _getChunk() {
    final length = _totalChunks == 1 ? _fileSize : chunkByteSize;
    final start = length * _chunkCount;

    _chunk = file!.openRead(start, start + length);
    if (start + length <= _fileSize)
      _chunkLength = length;
    else
      _chunkLength = _fileSize - start;
  }

  /// Called on net failure. If retry [_attemptCount] < [attempts], retry after [delayBeforeAttempt]
  _manageRetries() {
    if (_attemptCount < attempts) {
      _attemptCount = _attemptCount + 1;
      Timer(Duration(seconds: delayBeforeAttempt), () => _sendChunks());

      if (onAttemptFailure != null)
        onAttemptFailure!(
          'An error occurred uploading chunk $_chunkCount. ${attempts - _attemptCount} retries left.',
          _chunkCount,
          attempts - _attemptCount,
        );

      return;
    }

    _uploadFailed = true;

    if (onError != null)
      onError!(
        'An error occurred uploading chunk $_chunkCount. No more retries, stopping upload',
        _chunkCount,
        _attemptCount,
      );
  }

  /// Manages the whole upload by calling [_getChunk] and [_sendChunk]
  _sendChunks() async{
    if (_paused || _offline || _stopped)
      return;

   final presigned_url=await uploadUrlRequest(index: (_chunkCount+1).toString(), upload_id: upload_id, upload_key: upload_key, endPoint: endPoint, clientID: clientID, libraryID: libraryID, apiKey: apiKey );
   if(presigned_url!=""){
     await  _getChunk();
     await _sendChunk(presigned_url,_chunkLength).then((res) async{
       if (successfulChunkUploadCodes.contains(res.statusCode)) {
         chunk_list.add({"PartNumber":  _chunkCount+1, "ETag": res.headers.map['etag']!.toList().first});
         _chunkCount++;
         if (_chunkCount < _totalChunks) {
           _attemptCount = 0;
           _sendChunks();
         } else {
           final result=await videoUploadComplete(endPoint: endPoint, clientID: clientID, libraryID: libraryID, apiKey: apiKey, upload_id: upload_id, upload_key: upload_key, chunk_list: chunk_list);
           if (onSuccess != null && result.statusCode==200) onSuccess!();
         }

         if (onProgress != null) {
           double percentProgress = 100.0;
           if (_chunkCount < _totalChunks) {
             final bytesSent = _chunkCount * chunkByteSize;
             percentProgress = bytesSent * 100.0 / _fileSize;
           }
           onProgress!(percentProgress);
         }
       }
       else if (temporaryErrorCodes.contains(res.statusCode)) {
         if (_paused || _offline || _stopped)
           return;

         _manageRetries();
       }
       else {
         if (_paused || _offline || _stopped)
           return;

         _uploadFailed = true;

         if (onError != null)
           onError!(
             'Server responded with ${res.statusCode}. Stopping upload.',
             _chunkCount,
             _attemptCount,
           );
       }
     },
         onError: (err) {
           if (_paused || _offline || _stopped)
             return;

           // this type of error can happen after network disconnection on CORS setup
           _manageRetries();
         }
     );
   }

  }

  /// Restarts the upload after if the upload failed and came to a complete stop

  // restart() {
  //   if (!_uploadFailed)
  //     throw Exception('Upload hasn\'t yet failed, please use restart only after all retries have failed.');
  //
  //   _chunkCount = 0;
  //   chunkByteSize = chunkSize * 1024;
  //   _attemptCount = 0;
  //   _currentCancelToken = null;
  //
  //   _offline = false;
  //   _paused = false;
  //   _stopped = false;
  //   _uploadFailed = false;
  //
  //   _sendChunks();
  // }



}


lib/src/up_chunk_options.dart
// import 'dart:io';
//
// /// Defines the options to be passed to [UpChunk]'s [createUpload] method
// class UpChunkOptions {
//   /// Upload url as [String], required if [endPointResolver] == null
//   String? endPoint;
//
//   String? clientID;
//
//   String? libraryID;
//
//   String? apiKey;
//
//   String? videoID;
//
//   String? collectionID;
//
//
//
//   /// Future that returns the upload url as [String], required if [endPoint] == null
//   Future<String>? endPointResolver;
//
//   /// [File] to upload, required
//   File? file;
//
//   /// A Map with any headers you'd like included with the PUT request for each chunk.
// //  Map<String, String> headers = Map<String, String>();
//
//   /// The size in kb of the chunks to split the file into,
//   /// with the exception of the final chunk which may be smaller.
//   /// This parameter should be in multiples of 64
//   int chunkSize = 5120;
//
//   /// The number of times to retry any given chunk.
//   int attempts = 5;
//
//   /// Number of seconds to wait before a retry is fired
//   int delayBeforeAttempt = 1;
//
//   /// Fired when the client has gone online.
//   void Function()? onOnline;
//
//   /// Fired when the client has gone offline.
//   void Function()? onOffline;
//
//   /// Fired immediately before a chunk upload is attempted.
//   ///
//   /// [chunkNumber] is the number of the current chunk being attempted,
//   /// and [chunkSize] is the size (in bytes) of that chunk.
//   void Function(int chunkNumber, int chunkSize)? onAttempt;
//
//   /// Fired when an attempt to upload a chunk fails.
//   ///
//   /// [message] is the error message
//   /// [chunkNumber] is the number of the current chunk being attempted,
//   /// and [attemptsLeft] is the number of attempts left before stopping the upload
//   void Function(String message, int chunkNumber, int attemptsLeft)? onAttemptFailure;
//
//   /// Fired when a chunk has reached the max number of retries or the response code is fatal and implies that retries should not be attempted.
//   void Function(String message, int chunk, int attempts)? onError;
//
//   /// Fired when the upload is finished successfully.
//   void Function()? onSuccess;
//
//   /// Fired continuously with incremental upload progress. This returns the current percentage of the file that's been uploaded.
//   ///
//   /// [progress] a number from 0 to 100 representing the percentage of the file uploaded
//   void Function(double progress)? onProgress;
// }

lib/src/up_chunk.dart
import 'dart:async';
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:mime/mime.dart';
import 'package:http/http.dart' as http;
import 'connection_status_singleton.dart';
import 'uploader_repository.dart';


class GotipathUploader {
  /// HTTP response codes implying the PUT method has been successful
  final successfulChunkUploadCodes = const [200, 201, 202, 204, 308];

  /// HTTP response codes implying a chunk may be retried
  final temporaryErrorCodes = const [408, 502, 503, 504];

  String? endPoint;

  String? clientID;

  String? libraryID;

  String? apiKey;

  String? videoID;

  String? collectionID;

  String? upload_id;

  String? upload_key;

  List<Map<String,dynamic>> chunk_list=[];

  Future<String>? endPointResolver;
  File? file;
//  Map<String, String> headers = {};
  int chunkSize = 5120;
  int attempts = 5;
  int delayBeforeAttempt = 1;

  Stream<List<int>> _chunk = Stream.empty();
  int _chunkLength = 0;
  int _fileSize = 0;
  int _chunkCount = 0;
  int chunkByteSize = 0;
  String? _fileMimeType;
  Uri _endpointValue = Uri();
  int _totalChunks = 0;
  int _attemptCount = 0;
  bool _offline = false;
  bool _paused = false;
  bool _stopped = false;

  CancelToken? _currentCancelToken;

  bool _uploadFailed = false;

  void Function()? onOnline;
  void Function()? onOffline;
  void Function(int chunkNumber, int chunkSize)? onAttempt;
  void Function(String message, int chunkNumber, int attemptsLeft)? onAttemptFailure;
  void Function(String message, int chunk, int attempts)? onError;
  void Function()? onSuccess;
  void Function(double progress )? onProgress;

  GotipathUploader();

   createUpload() => _internal();

  /// Internal constructor used by [createUpload]
  _internal() {


    _validateOptions();

    chunkByteSize = chunkSize * 1024;



    videoUploadRequest(endPoint: endPoint, clientID: clientID, libraryID: libraryID, apiKey: apiKey, filename: file!.path.split('/').last, videoID: videoID)
      .then((value) async {
        upload_id=value['uploadId'];
        upload_key=value['key'];
        _fileSize = await file!.length();
        _totalChunks =  (_fileSize / chunkByteSize).ceil();

        await _getMimeType();
      })
      .then((_) => _sendChunks());

    // restart sync when back online
    // trigger events when offline/back online
    ConnectionStatusSingleton connectionStatus = ConnectionStatusSingleton();


    connectionStatus.connectionChange.listen(_connectionChanged);
  }

  /// It pauses the upload, the [_chunk] currently being uploaded will finish first before pausing the next [_chunk]
   pause() => _paused = true;



  /// It resumes the upload for the next [_chunk]
  resume() {
    if (!_paused) return;

    _paused = false;
    _sendChunks();
  }

  stop() {
    _stopped = true;
    _uploadFailed = true;
    _currentCancelToken!.cancel(Exception('Upload cancelled by the user'));

    if (onError != null)
      onError!(
        'Upload cancelled by the user.',
        _chunkCount,
        _attemptCount,
      );
  }

  /// It gets [file]'s mime type, if possible
  _getMimeType() async {
    try {
      _fileMimeType = lookupMimeType(file!.path);
    } catch (_) {
      _fileMimeType = null;
    }
  }

  /// It validates the passed options
  _validateOptions() {
    if (endPoint == null && endPointResolver == null)
      throw new Exception('either endPoint or endPointResolver must be defined');

    if (file == null)
      throw new Exception('file can''t be null');

    if (chunkSize <= 0 || chunkSize % 64 != 0)
      throw new Exception('chunkSize must be a positive number in multiples of 64');

    if (attempts <= 0)
      throw new Exception('retries must be a positive number');

    if (delayBeforeAttempt < 0)
      throw new Exception('delayBeforeAttempt must be a positive number');
  }




  /// Callback for [ConnectionStatusSingleton] to notify connection changes
  ///
  /// if the connection drops [_offline] is marked as true and upload us paused,
  /// if connection is restore [_offline] is marked as false and resumes the upload
  _connectionChanged(dynamic hasConnection) {
    print("this is connection changed called");
    if (hasConnection) {
      if (!_offline)
        return;

      _offline = false;

      if (onOnline != null) onOnline!();

      _sendChunks();
    }

    if (!hasConnection) {
      _offline = true;

      if (onOffline != null) onOffline!();
    }
  }

  /// Sends [_chunk] of the file with appropriate headers
  Future<Response> _sendChunk(String presignedUrl,int chunkLenght) async {

    var putHeaders =  {"Accept":"*/*","Content-Length": chunkLenght, "Content-Type": "binary/octet-stream"};

    print("this is video upload headers ${_chunkCount} ${putHeaders}");



    // if (_fileMimeType != null){
    //   putHeaders.putIfAbsent(Headers.contentTypeHeader, () => _fileMimeType!);
    // }
    // headers.forEach((key, value) => putHeaders.putIfAbsent(key, () => value));

    if (onAttempt != null)
      onAttempt!(_chunkCount, _chunkLength);

    _currentCancelToken = CancelToken();

   // Response<Response> response =await http.put(Uri.parse(presignedUrl), body: _chunk, headers: {"Accept":"*/*","Content-Length": chunkByteSize.toString()});

    final response=await Dio().request(
      presignedUrl,
      options: Options(
          headers: putHeaders,
          method: "PUT",
          // contentType:'video/mp4',
          followRedirects: false,
          validateStatus: (status) {
            return true;
          }
      ),
      data: _chunk,
      onSendProgress: (int sent, int total) {
        if (onProgress != null) {
          final bytesSent = _chunkCount * chunkByteSize;
          final percentProgress = (bytesSent + sent) * 100.0 / _fileSize;

          if (percentProgress < 100.0)
            onProgress!(percentProgress);
        }
      },
      cancelToken: _currentCancelToken,
    );
    print("this is video upload response ${response}");
    // returns future with http response
    return response;
  }

  /// Gets [_chunk] and [_chunkLength] for the portion of the file of x bytes corresponding to [chunkByteSize]
  _getChunk() {
    final length = _totalChunks == 1 ? _fileSize : chunkByteSize;
    final start = length * _chunkCount;

    _chunk = file!.openRead(start, start + length);
    if (start + length <= _fileSize)
      _chunkLength = length;
    else
      _chunkLength = _fileSize - start;
  }

  /// Called on net failure. If retry [_attemptCount] < [attempts], retry after [delayBeforeAttempt]
  _manageRetries() {
    if (_attemptCount < attempts) {
      _attemptCount = _attemptCount + 1;
      Timer(Duration(seconds: delayBeforeAttempt), () => _sendChunks());

      if (onAttemptFailure != null)
        onAttemptFailure!(
          'An error occurred uploading chunk $_chunkCount. ${attempts - _attemptCount} retries left.',
          _chunkCount,
          attempts - _attemptCount,
        );

      return;
    }

    _uploadFailed = true;

    if (onError != null)
      onError!(
        'An error occurred uploading chunk $_chunkCount. No more retries, stopping upload',
        _chunkCount,
        _attemptCount,
      );
  }

  /// Manages the whole upload by calling [_getChunk] and [_sendChunk]
  _sendChunks() async{
    print("called send chunks $_paused");
    if (_paused || _offline || _stopped)
      return;

   final presigned_url=await uploadUrlRequest(index: (_chunkCount+1).toString(), upload_id: upload_id, upload_key: upload_key, endPoint: endPoint, clientID: clientID, libraryID: libraryID, apiKey: apiKey );
   if(presigned_url!=""){
     await  _getChunk();
     await _sendChunk(presigned_url,_chunkLength).then((res) async{
       if (successfulChunkUploadCodes.contains(res.statusCode)) {

         print({"PartNumber":  _chunkCount+1, "ETag": res.headers.map['etag']!.toList().first});
         chunk_list.add({"PartNumber":  _chunkCount+1, "ETag": res.headers.map['etag']!.toList().first});
         _chunkCount++;
         if (_chunkCount < _totalChunks) {
           _attemptCount = 0;
           _sendChunks();
         } else {
           final result=await videoUploadComplete(endPoint: endPoint, clientID: clientID, libraryID: libraryID, apiKey: apiKey, upload_id: upload_id, upload_key: upload_key, chunk_list: chunk_list);
           if (onSuccess != null && result.statusCode==200) onSuccess!();
         }

         if (onProgress != null) {
           double percentProgress = 100.0;
           if (_chunkCount < _totalChunks) {
             final bytesSent = _chunkCount * chunkByteSize;
             percentProgress = bytesSent * 100.0 / _fileSize;
           }
           onProgress!(percentProgress);
         }
       }
       else if (temporaryErrorCodes.contains(res.statusCode)) {
         if (_paused || _offline || _stopped)
           return;

         _manageRetries();
       }
       else {
         if (_paused || _offline || _stopped)
           return;

         _uploadFailed = true;

         if (onError != null)
           onError!(
             'Server responded with ${res.statusCode}. Stopping upload.',
             _chunkCount,
             _attemptCount,
           );
       }
     },
         onError: (err) {
           if (_paused || _offline || _stopped)
             return;

           // this type of error can happen after network disconnection on CORS setup
           _manageRetries();
         }
     );
   }

  }

  /// Restarts the upload after if the upload failed and came to a complete stop
  restart() {
    if (!_uploadFailed)
      throw Exception('Upload hasn\'t yet failed, please use restart only after all retries have failed.');

    _chunkCount = 0;
    chunkByteSize = chunkSize * 1024;
    _attemptCount = 0;
    _currentCancelToken = null;

    _offline = false;
    _paused = false;
    _stopped = false;
    _uploadFailed = false;

    _sendChunks();
  }
}
